source/material_model/diffusion_dislocation.cc

```diff
@@ -0,0 +1,527 @@
+      for ( unsigned int i=0; i < x_comp.size(); ++i)
+        sum_composition += x_comp[i];
+
+      if (sum_composition >= 1.0)
+        {
+          volume_fractions[0] = 0.0;  //background mantle
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1]/sum_composition;
+        }
+      else
+        {
+          volume_fractions[0] = 1.0 - sum_composition; //background mantle
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1];
+        }
+      return volume_fractions;
+    }
+
+    template <int dim>
+    double
+    Diffusion_Dislocation<dim>::
+    average_value ( const std::vector<double> &composition,
+                    const std::vector<double> &parameter_values,
+                    const enum averaging_scheme &average_type) const
+    {
+      double averaged_parameter = 0.0;
+      std::vector<double> volume_fractions = compute_volume_fractions(composition);
+
+      switch (average_type)
+        {
+          case arithmetic:
+          {
+            for (unsigned int i=0; i< volume_fractions.size(); ++i)
+              averaged_parameter += volume_fractions[i]*parameter_values[i];
+            break;
+          }
+          case harmonic:
+          {
+            for (unsigned int i=0; i< volume_fractions.size(); ++i)
+              averaged_parameter += volume_fractions[i]/(parameter_values[i]);
+            averaged_parameter = 1.0/averaged_parameter;
+            break;
+          }
+          case geometric:
+          {
+            for (unsigned int i=0; i < volume_fractions.size(); ++i)
+              averaged_parameter += volume_fractions[i]*std::log(parameter_values[i]);
+            averaged_parameter = std::exp(averaged_parameter);
+            break;
+          }
+          case maximum_composition:
+          {
+            const unsigned int i = (unsigned int)(std::max_element( volume_fractions.begin(),
+                                                                    volume_fractions.end() )
+                                                  - volume_fractions.begin());
+            averaged_parameter = parameter_values[i];
+            break;
+          }
+          default:
+          {
+            AssertThrow( false, ExcNotImplemented() );
+            break;
+          }
+        }
+      return averaged_parameter;
+    }
+
+
+    template <int dim>
+    std::vector<double>
+    Diffusion_Dislocation<dim>::
+    calculate_viscosities ( const std::vector<double> &volume_fractions,
+                            const double &pressure,
+                            const double &temperature,
+                            const SymmetricTensor<2,dim> &strain_rate) const
+    {
+      double R = 8.31446; // J mol-1 K-1
+
+      // Viscosities
+      double e2inv = std::sqrt(std::pow(second_invariant(strain_rate),2) + std::pow(min_strain_rate,2)); // include minimum strain rate
```

Comment:
make it const. same for the various variables below: if you could make it const, make it const.